pkg/ruler/rulestore/bucket_client.go

```diff
@@ -0,0 +1,312 @@
+	}
+
+	perUserRuleGroupListMap := map[string]rules.RuleGroupList{}
+
+	for _, user := range users {
+		namespaces, err := b.listNamespacesForUser(ctx, user)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, namespace := range namespaces {
+			userBucket := newRuleStoreUserBucketClient(user, b.bucket)
+			nsEncoded := generateRuleObjectKey(namespace, "")
+			err = userBucket.Iter(ctx, nsEncoded, func(group string) error {
+				group = strings.TrimSuffix(strings.TrimPrefix(group, nsEncoded), delim)
+				decodedGroup, err := base64.URLEncoding.DecodeString(group)
+				if err != nil {
+					return err
+				}
+				perUserRuleGroupListMap[user] = append(perUserRuleGroupListMap[user], &rules.RuleGroupDesc{
+					User:      user,
+					Namespace: namespace,
+					Name:      string(decodedGroup),
+				})
+				return nil
+			})
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return perUserRuleGroupListMap, nil
+}
+
+// ListRuleGroupsForUserAndNamespace returns all the active rule groups for a user from given namespace.
+// If namespace is empty, groups from all namespaces are returned.
+func (b *BucketRuleStore) ListRuleGroupsForUserAndNamespace(ctx context.Context, userID string, namespace string) (rules.RuleGroupList, error) {
+	userBucket := newRuleStoreUserBucketClient(userID, b.bucket)
+
+	var namespaces []string
+	var err error
+
+	if namespace != "" {
+		namespaces = []string{namespace}
+	} else {
+		namespaces, err = b.listNamespacesForUser(ctx, userID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	groupList := rules.RuleGroupList{}
+
+	for _, namespace := range namespaces {
+		prefix := generateRuleObjectKey(namespace, "")
+		err = userBucket.Iter(ctx, prefix, func(group string) error {
+			group = strings.TrimSuffix(strings.TrimPrefix(group, prefix), delim)
+			decodedGroup, err := base64.URLEncoding.DecodeString(group)
+			if err != nil {
+				return err
+			}
+			groupList = append(groupList, &rules.RuleGroupDesc{
+				User:      userID,
+				Namespace: namespace,
+				Name:      string(decodedGroup),
+			})
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return groupList, nil
+}
+
+// LoadRuleGroups loads rules for each rule group in the map.
+// Parameter with groups to load *MUST* be coming from one of the List methods.
+// Reason is that some implementations don't do anything, since their List method already loads the rules.
+func (b *BucketRuleStore) LoadRuleGroups(ctx context.Context, groupsToLoad map[string]rules.RuleGroupList) error {
+	ch := make(chan *rules.RuleGroupDesc)
+
+	// Given we store one file per rule group. With this, we create a pool of workers that will
+	// download all rule groups in parallel. We limit the number of workers to avoid a
+	// particular user having too many rule groups rate limiting us with the object storage.
+	g, gCtx := errgroup.WithContext(ctx)
+	for i := 0; i < b.loadConcurrency; i++ {
+		g.Go(func() error {
+			for gr := range ch {
+				if gr == nil {
```

Comment:
Nit: It is actually bit strange to send `nil` to the channel. Perhaps we should just skip it in the `for` loop below.